plugins/resource_monitor_plugin/resource_monitor_plugin.cpp

```diff
@@ -0,0 +1,231 @@
+   }
+
+   // Start main thread
+   void plugin_startup() {
+      ilog("Creating and starting monitor thread");
+      monitor_thread = std::thread( [this] {
+         fc::set_os_thread_name( "resmon" ); // console_appender uses 9 chars for thread name reporting. 
+         monitor_task();
+      } );
+   }
+
+   // System is shutting down.
+   void plugin_shutdown() {
+      ilog("shutdown...");
+
+      // Signal the monior thread to shut down
+      done = true;
+
+      // Wait for the thread to end
+      monitor_thread.join();
+
+      ilog("exit shutdown");
+   }
+
+private:
+   std::thread               monitor_thread;
+   std::atomic_bool          done {false};
+
+   static constexpr uint32_t def_sleep_time_in_secs = 2;
+   static constexpr uint32_t def_used_space_threshold_in_percentage = 90;
+
+   uint32_t                  sleep_time_in_ms {def_sleep_time_in_secs * 1000}; 
+   uint32_t                  used_space_threshold_in_percentage {def_used_space_threshold_in_percentage};
+
+   struct filesystem_info {
+      dev_t           st_dev; // device id of file system containing "file_path"
+      uintmax_t       min_available_bytes; // minimum number of available bytes the file system must maintain
+      bfs::path       path_name;
+   };
+
+   // Stores file systems to be monitored. Duplicate
+   // file systems are not stored.
+   std::vector<filesystem_info> filesystems;
+
+   void monitor_task() {
+      ilog("Entering monitor_task");
+
+      while ( true ) {
+         if ( is_any_filesystem_space_below_threshold() ) {
+            appbase::app().quit(); // This will gracefully stop Nodeos
+            break;
+         }
+         
+         std::this_thread::sleep_for( std::chrono::milliseconds( sleep_time_in_ms ));
+
+         if ( done ) {
+            break;
+         }
+      }
+   }
+
+   bool is_any_filesystem_space_below_threshold() {
+      // Go over each monitored file system
+      for (auto& fs: filesystems) {
+         boost::system::error_code ec;
+         auto info = bfs::space(fs.path_name, ec);
+         if ( ec ) {
+            // As the system is running and this plugin is not a critical
+            // part of the system, we should not exit.
+            // Just report the failure and continue;
+            wlog( "Unable to get space info for ${path_name}: [code: ${ec}] ${message}",
+               ("path_name", fs.path_name.string())
+               ("ec", ec.value())
+               ("message", ec.message()));
+
+            continue;
+         }
+
+         if (info.available < fs.min_available_bytes) {
+            wlog("Space usage on ${path}'s file system exceeded configured threshold ${threshold}%, initiate shuting down!!! available: ${available}, Capacity: ${capacity}, min_available_bytes: ${min_available_bytes}", ("path", fs.path_name.string()) ("threshold", used_space_threshold_in_percentage) ("available", info.available) ("capacity", info.capacity) ("min_available_bytes", fs.min_available_bytes));
```

Comment:
This should be something like "initiating shut down", or simply "shutting down".